File: 2025/1/week1/다리 만들기_백준_2146/Solution.java
import java.io.*;
import java.util.*;

class Solution{
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static int n;
    static int[] di = new int[]{0,1,0,-1};
    static int[] dj = new int[]{1,0,-1,0};

    public static void main(String[] args) throws IOException{
        n = Integer.parseInt(br.readLine());
        int[][] world = new int[n][n];
        for(int i = 0; i< n ;i++){
            String line = br.readLine();
            for(int j = 0; j<n;j++){
                if(line.charAt(j*2) == '1')
                    world[i][j]=1;
            }
        }

        int[][] visited = new int[n][n];
        int visitFlag = 1;
        HashMap<Integer, Deque<int[]>> islandOutline = new HashMap<>();
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if(world[i][j] == 1){
                    visitFlag++;
                    Deque<int[]> deque = new ArrayDeque<>();
                    deque.add(new int[]{i,j});
                    visited[i][j]=visitFlag;
                    world[i][j]=visitFlag;
                    Deque<int[]> outLines = new ArrayDeque<>();
                    while(!deque.isEmpty()){
                        int[] pos = deque.pollFirst();
                        for (int k = 0; k < 4; k++) {
                            int ni = di[k] + pos[0];
                            int nj = dj[k] + pos[1];
                            if(0<= ni && ni<n && 0<= nj && nj<n && visited[ni][nj]!=visitFlag){
                                if(world[ni][nj] == 1){
                                    world[ni][nj] = visitFlag;
                                    deque.add(new int[]{ni,nj});
                                }
                                if(world[ni][nj] == 0){
                                    outLines.add(new int[]{ni,nj,0});
                                }
                            }
                        }
                    }
                    islandOutline.put(visitFlag, outLines);
                }
            }
        }

        int result = 20_000;
        visitFlag += 10_000;
        for (Map.Entry<Integer, Deque<int[]>> entry : islandOutline.entrySet()) {
            Deque<int[]> islandOutlineInfo =  entry.getValue();
            int islandId = entry.getKey();
            visitFlag++;
            SEARCH:
            while (! islandOutlineInfo.isEmpty()){
                int[] outLineInfo = islandOutlineInfo.pollFirst();
                for (int k = 0; k < 4; k++) {
                    int ni = di[k] + outLineInfo[0];
                    int nj = dj[k] + outLineInfo[1];
                    if(0<= ni && ni<n && 0<= nj && nj<n && visited[ni][nj]!=visitFlag){
                        if(world[ni][nj] == 0){
                            islandOutlineInfo.add(new int[]{ni,nj, outLineInfo[2]+1});
                            visited[ni][nj]=visitFlag;
                        }
                        else if(world[ni][nj] != islandId){
                            result = Math.min(result, outLineInfo[2]+1);
                            break SEARCH;
                        }
                    }
                }
            }
        }
        System.out.println(result);
    }
}

File: 2025/1/week1/다리 만들기_백준_2146/solution.py
import sys
from collections import deque
input = sys.stdin.readline
di = [0,1,0,-1]
dj = [1,0,-1,0]

n = int(input())
world = [list(map(int, input().split())) for _ in range(n)]

# 섬 구분
visited = [[0 for _ in range(n)] for _ in range(n)]
flag = 1
## 각 섬이 바다와 맞닿는 테두리 저장
islandOutline = {}
for i in range(n):
    for j in range(n):
        if world[i][j] == 1:
            flag+=1
            world[i][j] = flag
            queue = deque()
            queue.append([i,j])
            visited[i][j] = flag
            islandOutline[flag] = deque()
            while queue:
                pos = queue.popleft()
                for k in range(4):
                    ni = pos[0]+di[k]
                    nj = pos[1]+dj[k]
                    if 0 <= ni < n and 0 <= nj < n:
                        if visited[ni][nj] != flag:
                            if world[ni][nj] == 1:
                                visited[ni][nj] = world[ni][nj] = flag
                                queue.append([ni,nj])
                            if world[ni][nj] == 0:
                                islandOutline[flag].append([ni,nj, 0])

# 섬 구분 확인
# for i in range(n):
#     for j in range(n):
#         print(world[i][j], end=" ")
#     print()

# 모든 섬들을 출발지점으로 가장 짧은 다리 탐색
result = 20000
flag+=10000
for islandId, outline in islandOutline.items():
    flag+=1
    breakFlag = False
    while outline and not breakFlag:
        posi, posj, dis = outline.popleft()
        for k in range(4):
            ni = posi + di[k]
            nj = posj + dj[k]
            if 0<=ni<n and 0<=nj<n:
                if visited[ni][nj]!=flag:
                    if world[ni][nj] == 0:
                        visited[ni][nj]=flag
                        outline.append([ni,nj, dis+1])
                    elif world[ni][nj] != islandId:
                        result = min(result, dis+1)
                        breakFlag = True
                        break

print(result)